app/shared/design-system/components/primitives/auto-input.tsx

```diff
@@ -0,0 +1,205 @@
+	#ownerDocument = globalThis.document;
+	#canvasContext = globalThis.document?.createElement('canvas').getContext('2d');
+	#typeOfValue(value) {
+		if (/^([-+]?\d*\.?\d+)$/.test(value)) return 'number';
+		if (/^(%|cm|mm|in|px|pt|pc|em|ex|ch|rem|vw|vh|vmin|vmax|vb|vi|svw|svh|lvw|lvh|dvw|dvh|lh|rlh)$/.test(value)) return 'unit';
+		if (/^(\w+)$/.test(value)) return 'keyword';
+		if (/^(\w+\()/.test(value)) return 'function';
+		return '';
+	}
+	#measureEvent(value, event, target) {
+		const {fontSize, fontFamily, fontWeight, textAlign, paddingLeft, paddingRight} = getComputedStyle(target);
+		this.#canvasContext.font = `${fontWeight} ${fontSize} ${fontFamily}`;
+		const eventOffset = event.offsetX;
+		let charsOffset = 0;
+		let pixelPadding = 0
+		switch (textAlign) {
+			case 'left': {
+				pixelPadding += parseFloat(paddingLeft);
+				break;
+			}
+			case 'right': {
+				pixelPadding += target.clientWidth - this.#canvasContext.measureText(value).width - parseFloat(paddingRight);
+				break;
+			}
+		}
+		let pixelOffset = pixelPadding;
+		let valueOffset = -1;
+		const valueTokens = value.split(/([*/,()\s]?)([-+]?\d*\.?\d+)|([*/,()\s]+)/).filter(value => value).map(value => {
+			return {tokenType: this.#typeOfValue(value), tokenValue: value}
+		});
+		for (const [tokenIndex, tokenObject] of valueTokens.entries()) {
+			const {tokenValue} = tokenObject;
+			const tokenOffset = this.#canvasContext.measureText(tokenValue).width;
+			if ((eventOffset >= pixelOffset && eventOffset <= pixelOffset + tokenOffset)) {
+				valueOffset = tokenIndex;
+				break;
+			}
+			charsOffset += tokenValue.length;
+			pixelOffset += tokenOffset;
+		}
+		return {valueOffset, valueTokens, charsOffset, pixelOffset, pixelPadding};
+	}
+	// PUBLIC
+	handleEvent(event) {
+		const eventTarget = event.target;
+		const currentTarget = this.#activeTarget || eventTarget;
+		if (currentTarget?.nodeName !== 'INPUT' || currentTarget?.type !== 'text' || currentTarget?.readOnly !== false) return;
+		switch (event.type) {
+			case 'pointerover': case 'focusin': {
+				if (this.#pointerCapture === false) this.#pointerCapture = !this.#passiveEvents.forEach(eventName => this.#ownerDocument.addEventListener(eventName, this, false));
+				break;
+			}
+			case 'pointerout': case 'focusout': {
+				if (this.#pointerCapture === true) this.#pointerCapture = !!this.#passiveEvents.forEach(eventName => this.#ownerDocument.removeEventListener(eventName, this, false));
+				break;
+			}
+			case 'pointerdown': {
+				this.#activeTarget = currentTarget;
+				currentTarget.state = currentTarget.props;
+				const currentTargetCursor = currentTarget.style.cursor;
+				switch (currentTargetCursor) {
+					case this.#resizeCursor: {
+						this.#ownerDocument.documentElement.style.cursor = currentTargetCursor;
+						break;
+					}
+					case this.#pointerCursor: {
+						// dismiss if already active
+						if (this.#selectTarget) return this.#selectTarget.blur();
+						event.preventDefault?.();
```

Comment:
there is an `event` where preventDefault isn't a function or why the `?`?

Reply:
yeah, from reading code one doesn't expect the event to be a not event, very unexpected